book/src/04_3_i2c.md

```diff
@@ -2,16 +2,99 @@
+
+To use a peripheral sensor first you must get an instance of the it. The sensor is represented as a struct that contains it address and the i2c bus. This is done using traits defined in the `embedded-hal` crate. The struct is public as it needs to be accessible from outside this crate. It's fields are private. We also implement an instantiating method. Methods can also be public or private. This method needs to be accessible from outside, so it's labelled `pub`. Note that written this way, the sensor instance takes ownership of the i2c bus.
+
+```rust 
+#[derive(Debug)]
+pub struct IMC42670P<I2C> {
+    /// The concrete I²C device implementation.
+    i2c: I2C,
+
+    /// Device address
+    address: SlaveAddr,
+}
+
+// ...
+
+impl<I2C, E>IMC42670P<I2C>
+where
+    I2C: i2c::WriteRead<Error = E> + i2c::Write<Error = E>,
+{
+    /// Create a new instance of the IMC42670P.
+    pub fn new(i2c: I2C, address: SlaveAddr) -> Result<Self, E> {
+
+        let imc42670p = IMC42670P { i2c, address };
+
+        Ok(imc42670p)
+    }
+// ...
+```
+
+#### Representation of Registers 
+
+The sensor's registers are represented as enums. Each variant has the register's address as value. The type `Register` implements a method that exposes the variant's address.
+
+```Rust 
+#[derive(Clone, Copy)]
+pub enum Register {
+    WhoAmI = 0x75,
+}
+
+impl Register {
+    fn address(&self) -> u8 {
+        *self as u8
+    }
+}
+
+```
+
+#### read_register() and write_register()
+
+We define a read and a write method, based on methods provided by the `embedded-hal` crate. They serve as helpers for more specific methods and as an abstraction that is adapted to a sensor with 8bit registers. Note how the `read_register()` method is based on a write_read()` method. The reason for this lie in the characteristics of the i2c protocol: wW need to write a command over the i2c bus to be able to read bits in a register. Helper methods can remain private as they don't need to be accessible from outside this crate. 
+
+```rust
+impl<I2C, E>IMC42670P<I2C>
+where
+    I2C: i2c::WriteRead<Error = E> + i2c::Write<Error = E>,
+{    
+    //...
+    fn write_register(&mut self, register: Register, value: u8) -> Result<(), E> {
+        let byte = value as u8;
+        self.i2c
+            .write(self.address as u8, &[register.address(), byte])
+    }
+
+    fn read_register(&mut self, register: Register) -> Result<u8, E> {
+        let mut data = [0];
+        self.i2c
+            .write_read(self.address as u8, &[register.address()], &mut data)?;
+        Ok(u8::from_le_bytes(data))
+    }
+}
+```
+
+✅ Implement a public method that reads the `WHOAMI` register with the address `0x0F`. 
+✅ Optional: Implement further methods that add features to the driver. Check the [documentation](https://3cfeqx1hf82y3xcoull08ihx-wpengine.netdna-ssl.com/wp-content/uploads/2021/07/DS-000451-ICM-42670-P-v1.0.pdf) for the respective registers and their addresses. Some ideas:
```

Comment:
Hmm, that URL looks like it is in a CDN and I fear it will go away in the future. Can we host a copy of the data sheet in our repo? Or is there a more stable URL we can use - maybe at Digikey or Mouser? They sometimes host data sheets for products.